### DISCUSSION TITLE
Does PS 7.4.0 ignore `TabExpansion`? Breaking change vs 5.1 and 7.x < 7.4?

### DISCUSSION THREAD

**Discussion by nightroman** [PowerShell/PowerShell#20726] [Q&A] [answered]
The below custom (and fake) TabExpansion2 works fine in 5.1 and 7.x < 7.4 and does not work in 7.4
It uses the "legacy" function `TabExpansion` in order to get something (Apple and Banana here).

**Test 1 (working fine)**

- Open powershell 5.1 or pwsh 7.3.x, run this TabExpansion2.ps1
- Try to complete something, completions get Apple and Banana
- NB Note that the console title changes, indicating calls to `TabExpansion`

**Test 2 (not working)**

- Open pwsh 7.4, run this TabExpansion2.ps1
- Try to complete something, it's not the expected Apple and Banana
- NB Note that the console title does not change, indicating no calls to `TabExpansion`

Custom TabExpansion2.ps1:

```powershell
function global:TabExpansion2 {
    [CmdletBinding(DefaultParameterSetName = 'ScriptInputSet')]
    param(
        [Parameter(ParameterSetName = 'ScriptInputSet', Mandatory= $true, Position = 0)]
        [string]$inputScript,
        [Parameter(ParameterSetName = 'ScriptInputSet', Mandatory = $true, Position = 1)]
        [int]$cursorColumn,
        [Parameter(ParameterSetName = 'AstInputSet', Mandatory = $true, Position = 0)]
        [System.Management.Automation.Language.Ast]$ast,
        [Parameter(ParameterSetName = 'AstInputSet', Mandatory = $true, Position = 1)]
        [System.Management.Automation.Language.Token[]]$tokens,
        [Parameter(ParameterSetName = 'AstInputSet', Mandatory = $true, Position = 2)]
        [System.Management.Automation.Language.IScriptPosition]$positionOfCursor,
        [Parameter(ParameterSetName = 'ScriptInputSet', Position = 2)]
        [Parameter(ParameterSetName = 'AstInputSet', Position = 3)]
        [Hashtable]$options
    )

    # parse input
    if ($PSCmdlet.ParameterSetName -eq 'ScriptInputSet') {
        $_ = [System.Management.Automation.CommandCompletion]::MapStringInputToParsedInput($inputScript, $cursorColumn)
        $ast = $_.Item1
        $tokens = $_.Item2
        $positionOfCursor = $_.Item3
    }

    $results = @(
        'Apple'
        'Banana'
    )
    function TabExpansion {
        $Host.UI.RawUI.WindowTitle = [datetime]::Now.ToString()
        $results
    }
    [System.Management.Automation.CommandCompletion]::CompleteInput($ast, $tokens, $positionOfCursor, $null)
}
```

**Comment by MartinGC94** [comment]
I think it was removed here: https://github.com/PowerShell/PowerShell/pull/18337

**Reply by nightroman** [reply]
This is not mentioned in Breaking Changes here -- https://learn.microsoft.com/en-gb/powershell/scripting/whats-new/what-s-new-in-powershell-74?view=powershell-7.4

**Reply by nightroman** [reply]
Thus, it is not quite clear if this removal was intentional or not.

**Reply by MartinGC94** [reply]
You certainly have a valid point that it should have been mentioned in the release notes but there's no reason to pretend that we don't know whether or not it's an intentional breaking change. The linked PR explicitly calls out that it's breaking and the title and summary explicitly states that it's removing old legacy code so this is not something that simply slipped under the radar.

**Reply by EdgeDev01** [reply]
i would give it a new purpose, to expand tab sizes when adding tabs to tab groups

**Comment by kilasuit** [comment]
@theJasonHelmick can this be added for the InteractiveUX WG to discuss please.

**Reply by theJasonHelmick** [reply]
Yes -- thank you @kilasuit -- added

**Comment by theJasonHelmick** [comment, marked as answer]
Thank you @nightroman for reporting this issue.  The Interactive Working Group has reviewed this - and yes, this was an oversight and will be updated in the change log.
